src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs

```diff
@@ -644,7 +644,8 @@ protected bool TryGetValue(FieldsToFetch.FieldToFetch fieldToFetch, Document doc
 
             //_loadedDocuments.Clear(); - explicitly not clearing this, we want to cache this for the duration of the query
 
-            _loadedDocuments[document.Id ?? string.Empty] = document;
+            // we have to clone, because the `document` instance is updated on result
```

Comment:
Can you point the relevant code where we update it? I didn't expect that this will be modified

Reply:
This happens here:
https://github.com/ravendb/ravendb/blob/4b2db3760ef5ddd2595f0fa5634115a42c967999/src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs#L411-L433

We are modifying the _instance_ (and dispose the pointer)

Reply:
Once line below  we have:

```
document.IgnoreDispose = true;
```

Now we don't set it on the _cloned_ instance. Shouldn't we?

Reply:
Hm... I didn't notice that. The killer for me is that the _instance_ was modified by changing the `Data` properly.

Reply:
I guess we can create a new instance here, instead, would save the copy?

Reply:
Here? Wouldn't be better to create new `Document` instance in `ReturnProjection()`?

Reply:
Okay, I changed the implementation to take that into account. A bit more complex, but we avoid adding a `Clone` call if this isn't the same scenario

Reply:
I like current solution.